src/v/pandaproxy/schema_registry/avro.cc

```diff
@@ -54,24 +54,67 @@ namespace pandaproxy::schema_registry {
 
 namespace {
 
-bool check_compatible(avro::Node& reader, avro::Node& writer) {
+using avro_compatibility_result = raw_compatibility_result;
+
+avro_compatibility_result check_compatible(
+  avro::Node& reader, avro::Node& writer, std::filesystem::path p = {}) {
+    auto type_to_upper = [](avro::Type t) {
+        auto s = toString(t);
+        std::transform(s.begin(), s.end(), s.begin(), ::toupper);
+        return s;
+    };
+    avro_compatibility_result compat_result;
     if (reader.type() == writer.type()) {
-        // Do a quick check first
-        if (!writer.resolve(reader)) {
-            return false;
+        // Do some quick checks first
+        // These are detectable by the blunt `resolve` check below, but we want
+        // to extract as much error info as possible.
+        if (reader.type() == avro::Type::AVRO_ARRAY) {
+            compat_result.merge(check_compatible(
+              *reader.leafAt(0), *writer.leafAt(0), p / "items"));
+        } else if (reader.hasName() && reader.name() != writer.name()) {
+            // The Avro library doesn't fully support handling schema resolution
+            // with name aliases yet. While `equalOrAliasedBy` is available,
+            // `writer.resolve(reader)` doesn't take into account the aliases.
+            // Once Avro supports name alias resolution, we should only return a
+            // name_mismatch when
+            // `!writer.name().equalOrAliasedBy(reader.name())`, however, in the
+            // meantime it is best to return a more specific error.
```

Comment:
For this comment see the unit test in this commit as well. And this is the avro code of `NodeRecord::resolve(...)` for reference: https://github.com/redpanda-data/avro/blob/1410e79f9df61669c2d52f6d0643e6c35156e615/lang/c%2B%2B/impl/NodeImpl.cc#L126-L134

Fixing this upstream might just be a matter of changing that line to use `equalOrAliasedBy` instead of `operator==` on the names, but the Java implementation looks quite different and I am no Avro expert so I can't easily tell if there's more needed there. So I went with the "wait for upstream" approach for now.
https://github.com/redpanda-data/avro/blob/1410e79f9df61669c2d52f6d0643e6c35156e615/lang/java/avro/src/main/java/org/apache/avro/Resolver.java#L61-L63

Reply:
> Given that the current released version would let this pass

It wouldn't pass because we currently just check `writer.resolve(reader)` and that fails without alias support. So this is as strict as before but with a more specific error message.

We could let these pass by handling the name/alias check here, but then I think it would be more appropriate to add alias support in the `resolve` method linked above either in our avro fork or upstream avro.

Reply:
>it wouldn't pass 

Ah, ok cool. I misunderstood an older thread about this